### PR TITLE
Jetpack Search/Sync: ensure syncing of post meta and taxonomies that are indexed (#2)

### LINKED ISSUE #8663: Search: Index a whitelist of Post Meta for searching
Unlike for VIPs, we can't index all post meta for a site because there will be too many fields (we actually limit VIPs to 10k post meta keys also), so I propose we create a whitelist of post meta that can then be used for filtering and sorting.

For the whitelist we should do some of the top used plugins (pls add to list):
- woocommerce
- co authors plus
- wp job manager
- advanced custom fields (??? is there a way to whitelist some of this but not all?)
- bbPress
- BuddyPress (?)
- Polylang
- WPML
- the-events-calendar
- events-calendar-pro
- nextgen-gallery

In addition we should add a set of post meta that developers can use that they know will get indexed in a particular way. I'd suggest the following meta:
- jetpack_search_meta_1_date
- jetpack_search_meta_1_int
- jetpack_search_meta_1_float
- jetpack_search_meta_1_bool
- jetpack_search_meta_1_text
- jetpack_search_meta_1_keyword

Maybe have 10 of these fields: 1-10?

### LINKED ISSUE #15672: Search: sync post-meta/taxonomy indexing list into jetpack and document process to expand it
We have a hard coded list of post meta fields and custom taxonomies that we put into the search index. That list is in class.jetpack-sync-module-search.php on wp.com. We need to:
- [ ] sync the list to jetpack and likely refactor it significantly
- [ ] use this list to filter what taxonomies are available in the UI 
- [ ] add the list of post meta keys that can be filtered on (see https://github.com/Automattic/jetpack/issues/8663)
- [ ] document and create examples of how to request adding post-meta/taxonomies to the list (probably on jetpack.com/support/search, but maybe also in the UI).

### PR DESCRIPTION
**Pull request by bluefuton**
Our search index includes a fixed list of post meta keys and custom taxonomies in order to limit the number of fields in the index (see documentation https://jetpack.com/support/search/frequently-asked-questions/#troubleshoot-custom-tax).

We have so far not been syncing the list from wp.com into Jetpack which has had many downsides:
- Uncertainty if it worked in all cases (also lack of tests)
- No easy way to add new fields
- No visibility into what is indexed
- No way to double check that the config that the user has will work.
- No way to provide post meta faceting in search

This fixes all of that.

The class-search.php file will be replacing the old-style class.jetpack-sync-module-search.php that we are currently using on wp.com for the indexing. See https://github.com/Automattic/wpes-lib/blob/master/src/indices/jetpack-search-v2-index.php#L842

Fixes #15672
Fixes #8663

See #18870 for historical discussions.
See also pcNPJE-5Y-p2.

#### Does this pull request change what data or activity we track or use?

This auto expands the post meta that is synced for sites with Jetpack Instant Search enabled. This data can already be synced by users if they manually add it to allow lists. 

#### Testing instructions:

New unit tests are included. The unit tests are one of the more important pieces since I expect that the list of post meta and taxonomies will evolve over time and we want to make sure that it is trivial to add a new field.

Manual Instructions:

Spin up a new Jetpack powered site
Add a compatible Jetpack Search Plan
add search allowed meta to a post ( jetpack-search-meta0, entity_types, minimum_amount, etc )
Verify the Sync Queue is empty (data synced) 
Verify the meta is synced to the WP.com Cache

#### Proposed changelog entry for your changes:
* Search: adds explicit post meta keys and taxonomies to sync into the index. Particularly important for integrating with third-party plugins.

### CODE REVIEW

**Comment by gibrown** [comment]
Just adding @mdbitz and myself as reviewers for when this is ready. I think cherry-picking was likely the only way. Thanks for cleaning up my mess. :)

File: projects/plugins/jetpack/tests/php/sync/class-test-jetpack-sync-search.php [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+		parent::setUp();
+
+		// create a post.
+		$this->post_id = $this->factory->post->create();
+		$this->sender->do_sync();
+	}
+
+	public function test_module_is_enabled() {
+		$this->assertTrue( (bool) Modules::get_module( 'search' ) );
+		$this->assertTrue( \Jetpack::is_module_active( 'search' ) );
```
Comment (bluefuton) [inline projects/plugins/jetpack/tests/php/sync/class-test-jetpack-sync-search.php:34]: This is failing at the moment: 

```
1) Test_Jetpack_Sync_Search::test_module_is_enabled
Failed asserting that false is true.

/usr/local/src/jetpack-monorepo/projects/plugins/jetpack/tests/php/sync/class-test-jetpack-sync-search.php:34
```

**Comment by mdbitz** [comment]
Status :: I've updated the code base so tests all pass and updated the Search module so it doesn't get loaded by default. Instead it only is loaded if the Search package is present and Instant Search is enabled.

Next Steps are for some manual testing/verification that meta is persisted to the WP.com cache and then we can push this to Crew Review.

**Comment by mdbitz** [comment]
I've been thinking about the design/architecture of this diff and I'm currently thinking this shouldn't be a Module.  The Modules are more for defining hooks/actions that should be synced by setup of listeners.

In this case we only need to modify the allow list which doesn't have to tie into Modules.

I wonder if it makes sense to move this into the Search package. when the package loads up use the package classes to check if search is active and if so add the filter. seems simpler.

**Comment by mdbitz** [comment]
Verified that the allow list is updated when Instant Search is enabled and the meta are being synced to WP.com. However changes are needed on WP.com to allow the meta.

**Review by jeherve — CHANGES_REQUESTED** [review]
This is going to need a rebase now I'm afraid.

**Comment by mdbitz** [comment]
:nod: I also need to check with the search packaging if the code should be moved as well :)

File: projects/packages/sync/src/class-actions.php [hunk 2]
```diff
@@ -675,6 +675,36 @@ public static function initialize_woocommerce() {
...
+	}
+
+	/**
+	 * Add Search updates to Sync Filters.
+	 *
+	 * @access public
+	 * @static
+	 *
+	 * @param array $sync_modules The list of sync modules declared prior to this filter.
+	 * @return array A list of sync modules that now includes Woo's modules.
```
Comment (dereksmart) [inline projects/packages/sync/src/class-actions.php:701]: Woo's modules? Copy pasta?

File: projects/packages/sync/src/class-actions.php [hunk 3]
```diff
@@ -675,6 +675,36 @@ public static function initialize_woocommerce() {
...
+	 * Initializes sync for Instant Search.
+	 *
+	 * @access public
+	 * @static
+	 */
+	public static function initialize_search() {
+		if ( false === class_exists( 'Automattic\\Jetpack\\Search\\Module_Control' ) ) {
+			return;
+		}
+		$search_module = new \Automattic\Jetpack\Search\Module_Control();
```
Comment (jeherve) [inline projects/packages/sync/src/class-actions.php:688]: Quick question: if the Sync package relies on that class and method, do we need to add the Search package as a new dependency of the Sync package, or do we only rely on the check above?

**Review by jeherve — COMMENTED** [review]
This looks good to me; that's quite the list! I have one small question, below.

File: projects/packages/sync/src/class-actions.php (hunk 3 above)
Comment (mdbitz) [reply projects/packages/sync/src/class-actions.php:688]: We can't have Sync dependent on Search. Long term this code will move to the Search package. 

We only want to modify the allow lists if the site has an active Instant Search plan which is why we are checking for the class definition to determine if any loading should occur.

File: projects/packages/sync/src/class-actions.php (hunk 2 above)
Comment (mdbitz) [reply projects/packages/sync/src/class-actions.php]: 👍 thanks

**Review by jeherve — APPROVED** [review]

**Comment by anomiex** [comment]
This PR added **1300** new tests, that increases the time it takes to run Jetpack's normal PHPUnit test suite from [about 2 minutes](https://github.com/Automattic/jetpack/runs/4900723927?check_suite_focus=true#step:9:4708) to [**over 7**](https://github.com/Automattic/jetpack/runs/4901335282?check_suite_focus=true#step:9:4723).

Let's take another look at that, can we make it more efficient? Most of the problem seems to be that `WP_Test_Jetpack_Sync_Base::set_up()` is pretty slow.

**Pull request merged**